copt/randomized.py

```diff
@@ -691,3 +693,165 @@ def epoch_iteration_template(
+        f_deriv,
+        A,
+        b,
+        x0,
+        lmo,
+        step_size=None,
+        max_iter=500,
+        tol=1e-6,
+        verbose=False,
+        callback=None,
+        variant='SAGA'
+):
+    r"""Stochastic Frank-Wolfe (SFW) algorithm.
+
+    The SFW algorithm can solve optimization problems of the form
+
+        argmin_{x \in constraint} (1/n)\sum_{i}^n_samples f(A_i^T x, b_i)
+
+    Args:
+      f_deriv
+          derivative of f
+
+      x0: np.ndarray
+          Starting point for optimization.
+
+      step_size: function or None, optional
+          Step size for the optimization. If None is given, this will be set as the
+          default: t -> 2/(t+2)
+
+      lmo: function
+          returns the update direction
+
+      max_iter: int
+          Maximum number of gradient calls in the optimization.
+
+      tol: float
+          Tolerance criterion. The algorithm will stop whenever the
+          difference between two successive iterates is below tol.
+
+      verbose: bool
+          Verbosity level. True might print some messages.
+
+      callback: function or None
+          If not None, callback will be called at each iteration.
+
+      variant: str in {'SAG', 'MK', 'LF'}
+          Controls which variant of SFW to use.
+          'SAG' is described in [1],
+          'SAGA' is yet to be described.
+          'MK' is described in [2],
+          'LF' is described in [3].
+
+    Returns:
+      opt: OptimizeResult
+          The optimization result represented as a
+          ``scipy.optimize.OptimizeResult`` object. Important attributes are:
+          ``x`` the solution array, ``success`` a Boolean flag indicating if
+          the optimizer exited successfully and ``message`` which describes
+          the cause of the termination. See `scipy.optimize.OptimizeResult`
+          for a description of other attributes.
+
+     References:
+
+    .. [NDTELP2020] Negiar, Geoffrey, Dresdner, Gideon, Tsai Alicia, El Ghaoui, Laurent, Locatello, Francesco, and Pedregosa, Fabian.
+    `"Stochastic Frank-Wolfe for Constrained Finite-Sum Minimization" <https://arxiv.org/abs/2002.11860v2>` arxiv:2002.11860v2 (2020).
+
+    .. [MHK2018] Mokhtari, Aryan, Hassani, Hamed, and Karbassi, Amin `"Stochastic Conditional Gradient Methods:
+From Convex Minimization to Submodular Maximization" <https://arxiv.org/abs/1804.09554>`_, arxiv:1804.09554 (2018)
+
+    .. [LF2020] Lu, Haihao, and Freund, Robert `"Generalized Stochastic Frank-Wolfe Algorithm with Stochastic 'Substitute' Gradient for Structured Convex Optimization"
+    <https://arxiv.org/pdf/1806.05123.pdf>`_, Mathematical Programming (2020).
+
+
+    """
+    n_samples, n_features = A.shape
+    x = np.reshape(x0, n_features).astype(float)
+    assert x.shape == (n_features,)
+    A = sparse.csr_matrix(A).copy()
+
+    dual_var = np.zeros(n_samples)
+    grad_agg = np.zeros(n_features)
+
+    if variant == 'LF':
+        agg = utils.safe_sparse_dot(A, x)
```

Comment:
Or is `agg` the prediction vector? Agg = aggregation?

Reply:
Yes that's it. `agg` is their `sigma` prediction vector. The other methods (ours + Mokhtari) don't have that variable.

Reply:
See above